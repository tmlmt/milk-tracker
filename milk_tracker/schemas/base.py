import copy

# See https://stackoverflow.com/a/70277752/4258693
from collections.abc import Iterator
from contextlib import contextmanager
from typing import Any, Dict

from pydantic import BaseModel, ConfigDict


class UpdatableBaseModel(BaseModel):
    """BaseModel which can be updated partially.

    See https://stackoverflow.com/a/75451357/4258693

    """

    model_config = ConfigDict(validate_assignment=True)

    def update(self, **kwargs: Dict[str, Any]) -> None:
        """Update the model with multiple assignments."""
        self.__class__.validate(self.__dict__ | kwargs)
        self.__dict__.update(kwargs)

    @contextmanager
    def delay_validation(self) -> Iterator[None]:
        """Context manager which enables to direct reassign individual values.

        And have field- and model- validators to run as expected.

        Example:
        -------
        with myModel.delay_validation():
            myModel.a = 2
            myModel.b = 3

        """
        original_dict = copy.deepcopy(self.__dict__)

        self.__config__.validate_assignment = False  # type: ignore[attr-defined]
        try:
            yield
        finally:
            self.__config__.validate_assignment = True  # type: ignore[attr-defined]

        try:
            self.__class__.validate(self.__dict__)
        except:
            self.__dict__.update(original_dict)
            raise
