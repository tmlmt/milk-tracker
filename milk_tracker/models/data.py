from pathlib import Path
from typing import List

import pandas as pd
from utils.time_utils import timedelta_to_hrmin


class DataModel:
    """Holds the big data table as a Pandas DataFrame."""

    def __init__(self, file_path: Path) -> None:  # noqa: D107
        self.file_path: Path = file_path
        self.base_fields: List[str] = ["date", "start_time", "end_time"]
        self.df: pd.DataFrame = pd.read_excel(file_path)
        self.load()

    def load(self) -> None:
        """Load data from excel file."""
        self.df = pd.read_excel(self.file_path)
        self.clean()
        self.prepare()

    def clean(self) -> None:
        """Fill gaps and convert to datetime."""
        # When there's no end_time, make it equal to start time
        self.df.loc[self.df["end_time"] == "?", "end_time"] = self.df["start_time"]
        # Turn date and time columns to datetime
        self.df["date"] = pd.to_datetime(self.df["date"])

    def prepare(self) -> None:
        """Prepare dataset before first use."""
        self.df = self.compute_columns(self.df)

    def compute_columns(self, df: pd.DataFrame) -> pd.DataFrame:
        """Calculate other columns.

        Parameters
        ----------
        df : pd.DataFrame
            Unprocessed dataset only containing base fields

        Returns
        -------
        pd.DataFrame
            Dataset containing other required columns for analysis

        """
        # Combine 'Date' and 'Time' into a single datetime column
        # Convert to string first and combine, to circumvent errors
        # in converting excel format to datetime
        df["start_datetime"] = df.apply(
            lambda row: pd.to_datetime(str(row["date"].date()) + " " + str(row["start_time"])),
            axis=1,
        )
        df["end_datetime"] = df.apply(
            lambda row: pd.to_datetime(str(row["date"].date()) + " " + str(row["end_time"])),
            axis=1,
        )

        df.loc[df["end_datetime"] < df["start_datetime"], "end_datetime"] += pd.Timedelta(days=1)

        # Calculate duration
        df["duration"] = df["end_datetime"] - df["start_datetime"]
        df["duration_hrmin"] = df["duration"].apply(timedelta_to_hrmin)
        df["duration_min"] = round(df["duration"].dt.total_seconds() / 60, 2)

        # Calculate other things of interest
        df["previous_end_datetime"] = df["end_datetime"].shift(1)
        df["time_since_previous_start"] = df["start_datetime"].diff()
        df["time_since_previous_start_hrmin"] = df["time_since_previous_start"].apply(
            timedelta_to_hrmin
        )
        df["time_since_previous_start_hrs"] = round(
            df["time_since_previous_start"].dt.total_seconds() / 3600, 2
        )

        df["time_since_previous_end"] = df["start_datetime"] - df["previous_end_datetime"]
        df["time_since_previous_end_hrmin"] = df["time_since_previous_end"].apply(
            timedelta_to_hrmin
        )
        df["time_since_previous_end_hrs"] = round(
            df["time_since_previous_end"].dt.total_seconds() / 3600, 2
        )

        return df

    def add(self, date: str, start_time: str, end_time: str) -> None:
        """Add meal to dataset.

        Parameters
        ----------
        date : str
            Date YYYY-MM-DD
        start_time : str
            Start time HH:MM
        end_time : str
            End time HH:MM

        """
        new_meal = pd.DataFrame(
            [
                {
                    "date": pd.to_datetime(date),
                    "start_time": str(pd.to_datetime(start_time).time()),
                    "end_time": str(pd.to_datetime(end_time).time()),
                }
            ]
        )
        # To compute all columns, we also need the previous meal
        new_entry = self.compute_columns(
            pd.concat([self.df.iloc[[-1]][self.base_fields], new_meal])
        ).iloc[[-1]]
        # Merging and saving to file
        self.df = pd.concat([self.df, new_entry])

    def delete_latest(self) -> None:
        """Delete latest meal from dataset."""
        self.df = self.df.drop(self.df.tail(1).index)

    def save_to_file(self) -> None:
        """Save data back to excel file."""
        self.df[self.base_fields].to_excel(self.file_path, index=False)
