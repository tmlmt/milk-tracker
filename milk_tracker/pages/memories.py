from controllers.app import AppController
from nicegui import events, ui
from pages.header import display_header
from utils.time_utils import get_current_date


def page(mt: AppController) -> None:
    """Display Memories page.

    Parameters
    ----------
    mt : AppController
        Milk Tracker instance

    """
    display_header()
    ui.markdown("# Milk Tracker")
    ui.markdown("## Memories")
    mt.load_memories()

    table = ui.table.from_pandas(mt.memories.df).classes("w-[30rem]")
    table.add_slot(
        "header",
        r"""
        <q-tr :props="props">
            <q-th v-for="col in props.cols" :key="col.name" :props="props">
                {{ col.label }}
            </q-th>
            <q-th auto-width />
        </q-tr>
    """,
    )
    table.add_slot(
        "body",
        r"""
        <q-tr :props="props">
            <q-td key="date" :props="props">
                {{ props.row.date }}
                <q-popup-edit v-model="props.row.date" v-slot="scope"
                    @update:model-value="() => $parent.$emit('edit_entry', props.row)"
                >
                    <q-input v-model="scope.value"
                        dense autofocus counter
                        @keyup.enter="scope.set"
                    />
                </q-popup-edit>
            </q-td>
            <q-td key="description" :props="props">
                {{ props.row.description }}
                <q-popup-edit v-model="props.row.description" v-slot="scope"
                    @update:model-value="() => $parent.$emit('edit_entry', props.row)"
                >
                    <q-input v-model.number="scope.value" type="text"
                        dense autofocus counter
                        @keyup.enter="scope.set"
                    />
                </q-popup-edit>
            </q-td>
            <q-td auto-width >
                <q-btn size="sm" color="warning" round dense icon="delete"
                    @click="() => $parent.$emit('delete', props.row)"
                />
            </q-td>
        </q-tr>
    """,
    )

    def add_row() -> None:
        table.rows.append({"date": get_current_date(), "description": "New guy"})
        ui.notify("Added new row")
        table.update()

    def edit_entry(e: events.GenericEventArguments) -> None:
        for row in table.rows:
            if row["date"] == e.args["date"]:
                row.update(e.args)
        ui.notify(f"Updated rows to: {table.rows}")
        table.update()

    def delete(e: events.GenericEventArguments) -> None:
        table.rows[:] = [row for row in table.rows if row["id"] != e.args["id"]]
        ui.notify(f'Deleted row with ID {e.args["id"]}')
        table.update()

    with table.add_slot("bottom-row"):
        with table.cell().props("colspan=3"):
            ui.button("Add row", icon="add", color="accent", on_click=add_row).classes(
                "w-full"
            )

    table.on("edit_entry", edit_entry)
    table.on("delete", delete)
