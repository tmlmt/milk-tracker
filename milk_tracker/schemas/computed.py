from datetime import timedelta
from typing import Literal

from pydantic import Field, StringConstraints
from typing_extensions import Annotated
from utils.time_utils import get_current_time

from .base import UpdatableBaseModel


class ComputedValues(UpdatableBaseModel):
    """All computed values."""

    current_time: Annotated[str, StringConstraints(pattern=r"^\d{2}:\d{2}:\d{2}$")] = Field(
        default_factory=lambda: get_current_time(include_sec=True),
        description="Current time in format HH:MM:SS",
    )
    time_since_latest_end: timedelta = Field(
        default_factory=lambda: timedelta(0),
        description="Time since latest end of meal",
    )
    time_since_latest_start: timedelta = Field(
        default_factory=lambda: timedelta(0),
        description="Time since latest start of meal"
    )  # fmt: skip
    latest_meal_info: str = Field(
        default="No info",
        description="Latest meal information"
    )  # fmt: skip
    is_ongoing_meal: bool = Field(
        default=False,
        description="Lock state for input field of start of new meal"
    )  # fmt: skip
    is_ongoing_meal_paused: bool = Field(
        default=False,
        description="Whether a ongoing meal is paused"
    )  # fmt: skip
    is_ongoing_meal_buttontxt: Literal["Pause", "New round"] = Field(
        default="Pause",
        description="Button label for ongoing meal"
    )  # fmt: skip
    timer_meal_round: Annotated[str, StringConstraints(pattern=r"^(\d{2}:)?\d{2}:\d{2}$")] = Field(
        default="00:00",
        description="Timer for meal round in format HH:mm",
    )  # fmt: skip
    has_mother_taken_vitamins_today: bool = Field(
        default=False,
        description="Whether the mother has taken her vitamins today"
    )  # fmt: skip
    has_baby_taken_vitamins_today: bool = Field(
        default=False,
        description="Weather baby has taken their vitamins today"
    )  # fmt: skip
