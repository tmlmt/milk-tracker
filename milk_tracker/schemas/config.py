from pathlib import Path
from typing import Any, Dict

from pydantic import BaseModel, DirectoryPath, Field, field_validator, model_validator
from typing_extensions import Self


class Config(BaseModel):
    """Configuration Model."""

    ASSETS_DIR: DirectoryPath = Field(..., description="Directory path for assets")
    DATA_FILE_NAME: Path = Field(..., description="Journal file in Excel format")
    PLOTLY_DEFAULT_CONFIG: Dict[str, Any] = Field(
        ..., description="Default config value for Plotly plots"
    )

    @field_validator("DATA_FILE_NAME")
    @classmethod
    def check_file_extension(cls, v: Path) -> Path:  # noqa: D102
        if v.suffix != ".xlsx":
            msg = "FILE_NAME must have an .xlsx extension"
            raise ValueError(msg)
        return v

    @model_validator(mode="after")
    def file_exists(self) -> Self:  # noqa: D102
        full_path = Path(self.ASSETS_DIR) / self.DATA_FILE_NAME
        if not full_path.is_file():
            msg = f"{full_path} does not exist"
            raise ValueError(msg)
        return self
