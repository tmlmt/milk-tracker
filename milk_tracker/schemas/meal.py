import re
from datetime import datetime
from typing import List, Optional, Self

from pydantic import BaseModel, Field, field_validator, model_validator
from utils.time_utils import (
    force_full_time,
    get_current_date,
    get_current_time,
    is_same_minute,
    is_time_format,
)

from .base import UpdatableBaseModel


class Meal(BaseModel):
    """Base model for meals."""

    date: str = Field(
        default_factory=lambda: get_current_date(),
        description="Date in format YYYY-MM-DD"
    )  # fmt: skip
    start_time: str = Field(
        default_factory=lambda: get_current_time(include_sec=True),
        description="Start time in format HH:mm:ss",
    )
    end_time: Optional[str] = Field(
        default="",
        description="End time in format HH:mm:ss"
    )  # fmt: skip

    @field_validator("date")
    @classmethod
    def validate_date(cls, v: str) -> str:  # noqa: D102
        if not re.match(r"\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$", v):
            msg = "date must be in format YYYY-MM-DD"
            raise ValueError(msg)
        return v

    @field_validator("start_time")
    @classmethod
    def validate_start_time(cls, v: str) -> str:  # noqa: D102
        if not is_time_format(v):
            msg = "start_time must be in HH:mm[:ss]"
            raise ValueError(msg)
        return force_full_time(v)

    def get_start_datetime(self) -> datetime:
        """Return start date and time in datetime format.

        Returns
        -------
        datetime
            Start date and time

        """
        return datetime.strptime(f"{self.date} {self.start_time}", "%Y-%m-%d %H:%M:%S")


class MealRound(UpdatableBaseModel):
    """Round of meal."""

    start_datetime: datetime = Field(
        default_factory=lambda: datetime.now(),
        description="Start datetime of round"
    )  # fmt: skip
    end_datetime: Optional[datetime] = Field(default=None, description="End datetime of round")
    is_active: bool = Field(
        default=True,
        description="Whether the round is active"
    )  # fmt: skip

    @model_validator(mode="after")
    def check_end_time_or_active(self) -> Self:
        """Check mutual exclusivity of end_time and is_active."""
        if self.is_active and self.end_datetime:
            msg = "There must be no end_time if the round is active"
            raise ValueError(msg)
        if not self.is_active and not self.end_datetime:
            msg = "There must be an end_time if the round is no longer active"
            raise ValueError(msg)
        return self


class OngoingMeal(Meal, UpdatableBaseModel):
    """Meal that started at a certain date and start_time but without end_time yet.

    Raises
    ------
    ValueError
        If end_time is provided or not empty

    """

    rounds: List[MealRound] = Field(default_factory=list, description="List of meal rounds")
    "List of rounds"

    @field_validator("end_time")
    @classmethod
    def validate_empty_end_time(cls, v: Optional[str]) -> Optional[str]:  # noqa: D102
        if v is not None and v != "":
            msg = "end_time must not be provided or be empty for OngoingMeal"
            raise ValueError(msg)
        return v

    def pause(self) -> None:
        """Pause latest round and register end time."""
        if len(self.rounds) == 0:
            return
        latest_round = self.rounds[len(self.rounds) - 1]
        if latest_round.is_active:
            latest_round.update(end_datetime=datetime.now(), is_active=False)

    def start_new_round(self) -> None:
        """Start new round after pausing the latest one if necessary."""
        if any(r.is_active for r in self.rounds):
            self.pause()
        self.rounds.append(MealRound())

    @model_validator(mode="after")
    def auto_start_first_round(self) -> Self:
        """Make sure OngoingMeal starts with an active meal round."""
        if len(self.rounds) == 0:
            # If the meal was started within this minute, we start the
            # round at this very second, to avoid a surprising discrepanc
            # between start_time defined as %H:%M and start_datetime with %S
            first_round_start_datetime = self.get_start_datetime()
            if is_same_minute(first_round_start_datetime):
                first_round_start_datetime = first_round_start_datetime.replace(
                    second=datetime.now().second
                )
            self.rounds.append(MealRound(start_datetime=first_round_start_datetime))
        return self


class FinishedMeal(Meal):
    """Meal that started at a certain date and start_time and ended at end_time."""

    end_time: str = Field(..., description="End time in format HH:mm")

    @field_validator("end_time")
    @classmethod
    def validate_non_empty_end_time(cls, v: str) -> str:  # noqa: D102
        if not is_time_format(v):
            msg = "end_time must be in HH:mm[:ss]"
            raise ValueError(msg)
        return force_full_time(v)
