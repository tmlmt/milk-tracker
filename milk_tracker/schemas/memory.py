from datetime import date

import pandas as pd
from pydantic import BaseModel, Field


class Memory(BaseModel):
    """Base model for meals."""

    date_field: date = Field(default_factory=date.today, description="Date", alias="date")
    description: str = Field(default="", description="Description of the memory")

    def to_dataframe(self) -> pd.DataFrame:
        """Convert Memory into a Dataframe entry."""
        return pd.DataFrame(
            data={"date": [self.date_field], "description": [self.description]}
        )
