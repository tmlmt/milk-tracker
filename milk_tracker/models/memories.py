from pathlib import Path

import pandas as pd
from schemas.memory import Memory


class MemoriesDataModel:
    """Holds the big data table as a Pandas DataFrame."""

    def __init__(self, file_path: Path) -> None:  # noqa: D107
        self.file_path: Path = file_path
        self.load()

    def load(self) -> None:
        """Load data from excel file."""
        self.df = pd.read_csv(self.file_path)
        self.df["date"] = pd.to_datetime(self.df["date"]).dt.date

    def add(self, memory: Memory) -> None:
        """Add memory to dataset and order by date.

        Parameters
        ----------
        memory: Memory
            Memory to add to the dataset

        """
        self.df = pd.concat([self.df, memory.to_dataframe()], ignore_index=True)
        self.df = self.df.sort_values(by=["date"]).reset_index(drop=True)

    def save_to_file(self) -> None:
        """Save data back to excel file."""
        self.df.to_csv(self.file_path, index=False)
