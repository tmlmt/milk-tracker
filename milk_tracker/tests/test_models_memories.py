import datetime
from datetime import date
from pathlib import Path

import pytest
from inline_snapshot import snapshot
from models.memories import MemoriesDataModel
from schemas.memory import Memory

fixture_file_path = Path(__file__).parent.resolve() / "fixtures" / "memory_sample.csv"
fixture_birthday = date(2024, 1, 1)


@pytest.fixture()
def m() -> MemoriesDataModel:
    """Return initial model to be used for testing."""
    return MemoriesDataModel(file_path=fixture_file_path, birthday=fixture_birthday)


def test_init_memories(m: MemoriesDataModel) -> None:
    """Test constructor."""
    assert m.df.to_dict("records") == snapshot(
        [
            {
                "date": datetime.date(2024, 3, 1),
                "description": "Yes, it did",
                "index": 0,
                "age": "2 months",
            },
            {
                "date": datetime.date(2024, 2, 1),
                "description": "I can't believe it happened!",
                "index": 1,
                "age": "1 month",
            },
        ]
    )


def test_add_row(m: MemoriesDataModel) -> None:
    """Test add_row()."""
    new_m = Memory(date="2024-02-15", description="A new memory")
    m.add(new_m)
    assert m.df.to_dict("records") == snapshot(
        [
            {
                "date": datetime.date(2024, 3, 1),
                "description": "Yes, it did",
                "index": 0,
                "age": "2 months",
            },
            {
                "date": datetime.date(2024, 2, 15),
                "description": "A new memory",
                "index": 1,
                "age": "6 weeks 3 days",
            },
            {
                "date": datetime.date(2024, 2, 1),
                "description": "I can't believe it happened!",
                "index": 2,
                "age": "1 month",
            },
        ]
    )


def test_remove_row(m: MemoriesDataModel) -> None:
    """Test remove_row()."""
    m.remove(0)
    assert m.df.to_dict("records") == snapshot(
        [
            {
                "date": datetime.date(2024, 2, 1),
                "description": "I can't believe it happened!",
                "index": 0,
                "age": "1 month",
            }
        ]
    )


def test_edit_row(m: MemoriesDataModel) -> None:
    """Test edit_row()."""
    edited_m = Memory(date="2024-02-15", description="Edited memory")
    m.edit(0, edited_m)
    assert m.df.to_dict("records") == snapshot(
        [
            {
                "date": datetime.date(2024, 2, 15),
                "description": "Edited memory",
                "index": 0,
                "age": "6 weeks 3 days",
            },
            {
                "date": datetime.date(2024, 2, 1),
                "description": "I can't believe it happened!",
                "index": 1,
                "age": "1 month",
            },
        ]
    )
