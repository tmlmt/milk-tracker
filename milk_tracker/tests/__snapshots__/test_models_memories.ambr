# serializer version: 1
# name: test_add_row
  dict({
    'date': dict({
      0: datetime.date(2024, 2, 1),
      1: datetime.date(2024, 2, 15),
      2: datetime.date(2024, 3, 1),
    }),
    'description': dict({
      0: "I can't believe it happened!",
      1: 'A new memory',
      2: 'Yes, it did',
    }),
  })
# ---
# name: test_init_memories
  dict({
    'date': dict({
      0: datetime.date(2024, 2, 1),
      1: datetime.date(2024, 3, 1),
    }),
    'description': dict({
      0: "I can't believe it happened!",
      1: 'Yes, it did',
    }),
  })
# ---
